scvi/external/cellassign/_module.py

```diff
@@ -0,0 +1,188 @@
+from scvi.distributions import NegativeBinomial
+from torch.distributions import Dirichlet, Normal
+
+
+class CellAssignModule(BaseModuleClass):
+    """
+    Model for CellAssign.
+
+    Parameters
+    ----------
+    n_genes
+        Number of input genes
+    n_labels
+        Number of input cell types
+    rho
+        Binary matrix of cell type markers
+    **model_kwargs
+        Additional kwargs
+    """
+
+    def __init__(
+        self,
+        n_genes: int,
+        n_labels: int,
+        rho: torch.Tensor,
+        **model_kwargs,
+    ):
+        super().__init__()
+        self.n_genes = n_genes
+        self.n_labels = n_labels
+        self.rho = rho
+
+        self.register_buffer("cell_type_markers", rho)
+
+        # perform all other initialization
+        self.LOWER_BOUND = 1e-10
+        self.THETA_LOWER_BOUND = 1e-20
+        self.B = 10
+        self.min_delta = 2
+        self.dirichlet_concentration = [1e-2] * self.n_labels
+        self.shrinkage = True
+
+        # compute theta
+        self.theta_logit = torch.nn.Parameter(torch.randn(self.n_labels))
+
+        # compute delta (cell type specific overexpression parameter)
+        self.delta_log = torch.nn.Parameter(
+            torch.FloatTensor(self.n_genes, self.n_labels).uniform_(
+                np.log(self.min_delta), 2
+            )
+        )
+
+        # shrinkage prior on delta
+        if self.shrinkage:
+            self.delta_log_mean = torch.nn.Parameter(torch.Tensor(0))
+            self.delta_log_variance = torch.nn.Parameter(torch.Tensor(1))
+
+        self.log_a = torch.nn.Parameter(torch.zeros(self.B, dtype=torch.float64))
+
+    def _get_inference_input(self, tensors):
+        return {}
+
+    def _get_generative_input(self, tensors, inference_outputs):
+        x = tensors[_CONSTANTS.X_KEY]
+        y = tensors[_CONSTANTS.LABELS_KEY]
+
+        input_dict = dict(x=x, y=y)
+        return input_dict
+
+    @auto_move_data
+    def inference(self):
+        return {}
+
+    @auto_move_data
+    def generative(self, x, y):
+        # x has shape (n, g)
+        # n = 128
+        # g = 100
+        # c = 5
+        delta = torch.exp(self.delta_log)  # (g, c)
+        theta_log = F.log_softmax(self.theta_logit)  # (c)
+
+        # compute mean of NegBin - shape (n_cells, n_genes, n_labels)
+        s = x.sum(1, keepdim=True)  # (n, 1)
+        base_mean = torch.log(s)
+        base_mean_u = base_mean.unsqueeze(-1)  # (n, 1, 1)
+        base_mean_e = base_mean_u.expand(
+            s.shape[0], self.n_genes, self.n_labels
+        )  # (n, g, c)
+
+        delta_rho = delta * self.rho
+        delta_rho_e = delta_rho.expand(
+            s.shape[0], self.n_genes, self.n_labels
+        )  # (n, g, c)
+        log_mu_ngc = base_mean_e + delta_rho_e
+        mu_ngc = torch.exp(log_mu_ngc)  # (n, g, c)
+
+        # compute basis means for phi - shape (B)
+        basis_means_fixed = np.linspace(torch.min(x), torch.max(x), self.B)
```

Comment:
let's use torch linspace and use the device argument of linspace to be the device of `x`, this allows us to write code that can be used on gpu or cpu

Reply:
basically in case I miss it, anywhere you create a torch Tensor in this particular function, you need to specify the device using the device of the inputs

Reply:
and regarding `basis_means`, below we should also use torch functions everywhere